### ISSUE TITLE
Add tf.variable_scope to all sym operations in the TF tree

### ISSUE THREAD

**Issue by ryanjulian** [ryanjulian/rllab#43]
This will make the TF graph display much easier to read, with no user effort.

**Comment by ryanjulian** [comment]
A sym operation is a symbolic operation (i.e. it uses TensorFlow operations on the abstract computation graph, not numpy or Python operations on concrete numbers).

For instance, GaussianMLPPolicy.dist_info_sym() is a symbolic operation.

tf.variable_scope() tells TensorFlow that a series of operations is named in a particular way. It helps make the computation graph easier to read by grouping the related graph nodes.

This bug is to add a tf.variable_scope() to all symbolic operations in primitives like distributions, policies, etc. The scope name should be prepended with the name of the primitive (e.g. primitive_name/op_name), because scopes must be unique.

**Comment by ghost** [comment]
I'm a little bit confused about the outcome of using the variable scope feature. I ran the cartpole training and obtained the following graph in TensorBoard.
![graph_cartpole](https://user-images.githubusercontent.com/35857569/40092668-59031de4-5873-11e8-81a4-2da703346915.png)
The GaussianMLPPolicy already implements variable scope for its constructor, and in this example the scope name is just policy. The expanded view of policy variable scope is the following:
![graph_cartpole_policy](https://user-images.githubusercontent.com/35857569/40092688-7c3a3162-5873-11e8-9c9b-343cc9acfad0.png)
Method dist_info_sym(), a symbolic operation in class GaussianMLPPolicy, is called both in the GaussianMLPPolicy constructor and outside the class. By adding a variable scope in dist_info_sym() all the nodes in dist_info_sym are now encapsulated within the new policy variable scope:
![graph_cartpole_policy_3](https://user-images.githubusercontent.com/35857569/40092705-9b3184b2-5873-11e8-95b2-d57cde7cc9e5.png)
And the nodes created by an external call are placed in the policy_1:
![graph_cartpole_policy_2](https://user-images.githubusercontent.com/35857569/40092710-a683732a-5873-11e8-8d9a-cadaf6f9e93b.png)
I just want to confirm if this is what is expected by implementing variable scope.

**Comment by ryanjulian** [comment]
Can you send me the TF graph file on Slack?

**Comment by ryanjulian** [comment]
Oh I see, the difference. TF creates a new scope `policy_1` with a counter when `dist_info_sym()` is called from outside `policy` and the scope is reused.

tl;dr: no that is not what we intended.

There is a setting in TF for reusing scopes but I'm not sure that's exactly what we want (but it might be required to implement what we want).

Different scopes for different operations should be unique. I think the important question is "what should do we want the computation graph to look like for this case?"

Here are some disorganized thoughts:
* When I am debugging my program, I want all operations associated with the `policy` object to appear inside the `policy` subgraph. In this way, the computation graph resembles the inheritance/call graph of the Python program.
* To restate in general: we want the Python call graph to resemble as closely as possible the TF compute graph
* Different invocations of `dist_info_sym()` might have different inputs and outputs, so they can't really share an inner scope (scope names must be unique).
* Different invocations associated with a single policy should still appear in that policy's subgraph, but they need different names.

Here's one way to achieve that:
* Every TF-backed primitive object has a name for its enclosing scope. It should have a constructor parameter so I can set it to something special, but a sane default is fine. If someone reuses the same name (e.g. by invoking 2 primitives with out giving them special names), IIRC TensorFlow will do what it did above (make a copy and add a counter). It will be ugly in the compute graph but the program will not break.
* All operations called on that primitive object should be inside the scope of the primitive object (this is what is violated in your example).
* The above also applies to nested primitives with "has a" relationships: If a `GaussianMLPPolicy` object named `foo` has a `DiagonalGaussian` object named `bar`, then all operations on that particular `DiagonalGaussian` should appear in the scope `foo/bar`
* Because symbolic operations *also* give rise to new scopes/subgraphs, which must also be unique names. The logic from the primitive then applies recursively: they should accept an optional name parameter with a sane default. e.g. `GaussianMLPPolicy.dist_info_sym()` would appear in `policy/dist_info_sym` by default, or perhaps `foo/baz` if the the policy gets the optional parameter `name=foo` and `dist_info_sym()` gets the optional parameter `name=baz`.
* Last thought: it might also make it more readable to put the class/operation names inside the scope paths as well, regardless of naming. e.g. above `foo/bar` becomes `GaussianMLPPolicy[foo]/DiagonalGaussian[bar]`. To decide we would just have to try it and see how it looks on TensorBoard. For instance, perhaps that takes up too much space and should be omitted, or maybe we should make it `foo(GaussianMLPPolicy)` instead, etc. The aesthetics really depend on how it looks in TensorBoard.

**Comment by eric-heiden** [comment]
I agree with Ryan's points.
Just to clarify: we should probably add an optional `name` argument to the policy function `dist_info_sym`. This name gets appended to the policy's name, e.g. `policy/compute_kl` where "policy" is the name of the policy and "compute_kl" is the name provided to `dist_info_sym`. The generated vars should appear inside the policy's scope.

**Comment by eric-heiden** [comment]
It turns out to be difficult to make sure all operations on a primitive are bound to that primitive's scope without resulting in deeply nested scopes. For example, when you call `dist_info_sym` on the policy from within the `init_opt` function in `BathPolOpt`, the stack of scopes you're in when you call `dist_info_sym` gets copied inside the policy's primitive. `dist_info_sym` gets called in `BatchPolOpt` to reparameterize the policy for the new inputs, whereas `dist_info_sym` also gets called in the policy's constructor to get access to the mean and log_std output. So instead of trying to weld both functions into the same scope and causing lots of nesting, I think the graph is readable enough when the `dist_info_sym` calls are named properly and appear inside the scope of the operation/class that calls them.

**Comment by ryanjulian** [comment]
There is likely some escape hatch in the TensorFlow API to escape your
enclosing scopes. I'll try to take a look for it (currently on the road).

On Fri, May 18, 2018, 2:35 PM Eric Heiden <notifications@github.com> wrote:

> It turns out to be difficult to make sure all operations on a primitive
> are bound to that primitive's scope without resulting in deeply nested
> scopes. For example, when you call dist_info_sym on the policy from
> within the init_opt function in BathPolOpt, the stack of scopes you're in
> when you call dist_info_sym gets copied inside the policy primitive.
> dist_info_sym get called in BatchPolOpt to reparameterize the policy for
> the new inputs, whereas dist_info_sym also gets called in the policy's
> constructor to get access to the mean and log_std output. So I instead of
> trying to weld both functions into the same scope and causing lots of
> nesting, the graph is readable enough when the dist_info_sym calls are
> named properly and appear inside the scope of the operation/class that
> calls them.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ryanjulian_rllab_issues_43-23issuecomment-2D390338660&d=DwMFaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=4dbXvIJHO1nJKOCTF2eVfpb-CT8Bw831y709jar5haQ&s=WaOku-RuiLNnXiP0aIrkk4wj4wARO__V-Xb20lVZ7-c&e=>,
> or mute the thread
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABT-5F25uyRLEmkegRfly-5FJ34BWt1CtNTGks5tzz6PgaJpZM4TFY4D&d=DwMFaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=4dbXvIJHO1nJKOCTF2eVfpb-CT8Bw831y709jar5haQ&s=WE-alwlyRNtCRsendjX0DD5Fbu2A10Ap4RqjHnOvTnc&e=>
> .
>
-- 
PhD Student
USC CLMC/RESL
(510) 859-3224
rjulian@usc.edu

**Comment by ryanjulian** [comment]
https://www.tensorflow.org/programmers_guide/variables

Deeply nested call chains was always going to happen with this design but I
don't think it's a deal breaker.

See the form of tf.variable_scope(scope, reuse=False) at the end of this
article. Could we have each primitive store it's scope and then always pass
parent scope to the child scopes?

On Fri, May 18, 2018, 2:45 PM Ryan Julian <rjulian@usc.edu> wrote:

> There is likely some escape hatch in the TensorFlow API to escape your
> enclosing scopes. I'll try to take a look for it (currently on the road).
>
> On Fri, May 18, 2018, 2:35 PM Eric Heiden <notifications@github.com>
> wrote:
>
>> It turns out to be difficult to make sure all operations on a primitive
>> are bound to that primitive's scope without resulting in deeply nested
>> scopes. For example, when you call dist_info_sym on the policy from
>> within the init_opt function in BathPolOpt, the stack of scopes you're
>> in when you call dist_info_sym gets copied inside the policy primitive.
>> dist_info_sym get called in BatchPolOpt to reparameterize the policy for
>> the new inputs, whereas dist_info_sym also gets called in the policy's
>> constructor to get access to the mean and log_std output. So I instead of
>> trying to weld both functions into the same scope and causing lots of
>> nesting, the graph is readable enough when the dist_info_sym calls are
>> named properly and appear inside the scope of the operation/class that
>> calls them.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ryanjulian_rllab_issues_43-23issuecomment-2D390338660&d=DwMFaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=4dbXvIJHO1nJKOCTF2eVfpb-CT8Bw831y709jar5haQ&s=WaOku-RuiLNnXiP0aIrkk4wj4wARO__V-Xb20lVZ7-c&e=>,
>> or mute the thread
>> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_ABT-5F25uyRLEmkegRfly-5FJ34BWt1CtNTGks5tzz6PgaJpZM4TFY4D&d=DwMFaQ&c=clK7kQUTWtAVEOVIgvi0NU5BOUHhpN0H8p7CSfnc_gI&r=VEuK-fS6-fUHjpi4P-BnNQ&m=4dbXvIJHO1nJKOCTF2eVfpb-CT8Bw831y709jar5haQ&s=WE-alwlyRNtCRsendjX0DD5Fbu2A10Ap4RqjHnOvTnc&e=>
>> .
>>
> --
> PhD Student
> USC CLMC/RESL
> (510) 859-3224
> rjulian@usc.edu
>
-- 
PhD Student
USC CLMC/RESL
(510) 859-3224
rjulian@usc.edu

**Comment by ghost** [comment]
@ryanjulian please review my changes in #72 to follow up this issue.

**Comment by ryanjulian** [comment]
Fixed in #72
